lib/spack/spack/solver/counter.py

```diff
@@ -24,7 +25,13 @@ class Counter:
     """
 
     def __init__(self, specs: List["spack.spec.Spec"], tests: bool) -> None:
-        self.specs = specs
+        runtime_pkgs = spack.repo.PATH.packages_with_tags("runtime")
+        runtime_virtuals = set()
+        for x in runtime_pkgs:
+            pkg_class = spack.repo.PATH.get_pkg_class(x)
+            runtime_virtuals.update(pkg_class.provided_virtual_names())
+
+        self.specs = specs + [spack.spec.Spec(x) for x in runtime_pkgs]
```

Comment:
This requires too much effort from the packager. They shouldn't have to tag runtimes.

I'm assuming the reason we do this is because we need the full set of possible virtuals before the solve, but you *have* that after you call `runtime_constraints()` on all the packages that have that method. You can just see if any of the global dependencies added with `pkg("*").depends_on()` are virtual, and get the same (actually, possibly smaller) list of `runtime_virtuals` that you have here.  So this logic should go away and we should just look at what was injected.

Reply:
I can remove the tags for an extra method in e.g. `gcc` that returns what the package could possibly inject.